contributors/design-proposals/secure_watch.md

```diff
@@ -0,0 +1,208 @@
+# Background
+
+ As part of increasing security of a cluster, we are planning to limit the ability
+of a given Kubelet (in general: node), to be able to read only resources associated
+with it. Those resources, in particular means: secrets, configmaps &
+persistentvolumeclaims. This is needed to avoid situation when compromising node
+de facto means compromising a cluster. For more details & discussions see
+https://github.com/kubernetes/kubernetes/issues/40476.
+ 
+ However, by some extension to this effort, we would like to improve scalability
+of the system, by significantly reducing amount of api calls coming from kubelets.
+As of now, to avoid situation that kubelet is watching all secrets/configmaps/...
+in the system, it is not using watch. Instead of that, it is retrieving indidual pods,
+by sending individual GET requests. However, it is sending those requests periodically
+to enable automatic updates of mounted secrets/configmaps/... In large clusters,
+this is generating huge unnecessary load, as this load in principle should be
+watch-based. We would like to address this together with solving the authorization issue.
+
+# Proposal
+
+ In this proposal, I'm not focusing on how exactly security should be done - I'm just
+sketching very high level approach and exact authorization mechanism should be discussed
+separately.
+
+ At the high level, what we want to achieve is to enable LIST and WATCH requests to
+support filtering of "objects only attached to pods bound to a given node". We obviously
+want to be able to authorize other types of requests (in particular GETs), so that the
+design has to be consistent with that.
+
+ To solve this, I propose to introduce a new filtering mechanism (next to label selector
+and field selector): ```node selector ``` (we probably need better name though). Its
+semantic will be to filter only objects that are attached to pods bound to a given node
+and it will be supported for some predefined set of object types.
+
+# Detailed design
+
+ We will introduce the following ```node selector ``` filtering mechanism:
```

Comment:
@lavalamp - thanks for the suggestion. However, I think there are few things we need to think about it in this solution. The main thing is:
"are we going to reflect this information in etcd"

If not (e.g. it will be just in-memory), then it might be tricky to synchronize it between apiservers in HA setups.

If it will be in etcd, then:
1. having "allowed kubelets" be listed together with the object doesn't seem like a good idea
2. if we are not going to store the "newly allowed/disallowed kubelet" together with the "contentless update", then how we are going to derive its purpose in apiserver?

Maybe these questions are stupid, but I guess I'm not following your thoughts here.